src/main/scala/abalone/variant/Variant.scala

```diff
@@ -102,17 +102,88 @@ abstract class Variant private[variant] (
+      val activePlayerPieces = situation.board.piecesOf(situation.player)
+      val opponentPieces = situation.board.piecesOf(!situation.player)
+      val moves = List(
+        neighbour.dir(direction).toList.flatMap {
+          case (thirdSquareInLine) => {
+            if (situation.board.isEmptySquare(Some(thirdSquareInLine))) // xx.
+              generateMove(pos, thirdSquareInLine, "line")
+            else if (opponentPieces.contains(thirdSquareInLine)) // xxo
+              thirdSquareInLine.dir(direction) match { // xxo?
+                case None => generateMove(pos, thirdSquareInLine, "pushout") // xxo\
+                case Some(emptySquare) if situation.board.isEmptySquare(Some(emptySquare)) => {
+                  generateMove(pos, thirdSquareInLine, "push") // xxo.
+                }
+                case _ => None
+              }
+            else if (activePlayerPieces.contains(thirdSquareInLine)) // xxx
+              thirdSquareInLine.dir(direction).flatMap {
+                case (fourthSquareInLine) => // xxx_
+                  if (situation.board.isEmptySquare(Some(fourthSquareInLine))) // xxx.
+                    generateMove(pos, fourthSquareInLine, "line")
+                  else if (opponentPieces.contains(fourthSquareInLine)) // xxxo
+                    fourthSquareInLine.dir(direction) match { // xxxo?
+                      case None => generateMove(pos, fourthSquareInLine, "pushout") // xxxo\
+                      case Some(emptyPos) if (situation.board.isEmptySquare(Some(emptyPos))) => generateMove(pos, fourthSquareInLine, "push") // xxxo.
+                      case _ => fourthSquareInLine.dir(direction).flatMap { // xxxo?
+                        case (fifthSquareInLine) =>
+                          if (opponentPieces.contains(fifthSquareInLine)) // xxxoo
+                            fifthSquareInLine.dir(direction) match {
+                              case None => generateMove(pos, fourthSquareInLine, "pushout") // xxxoo\
+                              case Some(emptySquare) if situation.board.isEmptySquare(Some(emptySquare)) => generateMove(pos, emptySquare, "push") // xxxoo.
+                              case _ => None
+                            }
+                          else None
+                      }
+                    }
+                  else None
+              }
+            else None
+          }
+        }
+      )
 
-  def validPushes(@nowarn situation: Situation):  Map[Pos, List[Move]] = {
-    Map()
+      moves.flatten ++ generateSideMoves(
+        List(Some(pos), Some(neighbour), neighbour.dir(direction).flatMap{x => if (activePlayerPieces.contains(x)) Some(x) else None}).flatten,
+        direction
+      )
+    }
+
+    turnPieces(situation).flatMap {
+      case (pos, _) =>
+        Map(pos -> pos.neighbours.collect {
+          case Some(neighbour) if situation.board.piecesOf(situation.player).contains(neighbour) => (pos, neighbour, pos.dir(neighbour))
```

Comment:
will swap `situation.board.piecesOf(situation.player).` with the new `activePlayerPieces`